### PR TITLE
ovl_Magic_Dark

### PR DESCRIPTION
**Pull request by Sirius902**
Pog finally solved the mysteries of the dark magic. There was some funky stuff going on there.

### CODE REVIEW

File: src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c [hunk 1]
```diff
@@ -66,37 +69,218 @@ void MagicDark_Destroy(Actor* thisx, GlobalContext* globalCtx) {
...
+    thisx->scale.x *= 1.3f;
+    thisx->scale.z *= 1.3f;
+
+    phi_a0 = (this->timer < 20) ? (this->timer * 12) : 255;
+
+    if (nayrusLoveTimer >= 1180) {
+        this->primAlpha = 15595 - (nayrusLoveTimer * 13);
+        if (nayrusLoveTimer & 1) {
+            temp2 = this->primAlpha;
+            this->primAlpha = (temp2 & 0xFF) >> 1;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c:115]: Any chance this works?
```suggestion
            this->primAlpha >>= 1;
```

File: src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c [hunk 2]
```diff
@@ -66,37 +69,218 @@ void MagicDark_Destroy(Actor* thisx, GlobalContext* globalCtx) {
...
+        }
+    } else {
+        if (nayrusLoveTimer >= 1100) {
+            temp = (nayrusLoveTimer << 7);
+            temp += 127;
+            this->primAlpha = temp;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c:121]: Also, maybe this works?
```suggestion
            this->primAlpha = (u8)(nayrusLoveTimer << 7) + 127;
```

File: src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c [hunk 3]
```diff
@@ -66,37 +69,218 @@ void MagicDark_Destroy(Actor* thisx, GlobalContext* globalCtx) {
...
+    phi_a0 = (this->timer < 20) ? (this->timer * 12) : 255;
+
+    if (nayrusLoveTimer >= 1180) {
+        this->primAlpha = 15595 - (nayrusLoveTimer * 13);
+        if (nayrusLoveTimer & 1) {
+            temp2 = this->primAlpha;
+            this->primAlpha = (temp2 & 0xFF) >> 1;
+        }
+    } else {
+        if (nayrusLoveTimer >= 1100) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c:118]: else if here imo

File: src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c [hunk 4]
```diff
@@ -12,14 +12,13 @@
...
-void MagicDark_Draw(Actor* thisx, GlobalContext* globalCtx);
+void MagicDark_OrbUpdate(Actor* thisx, GlobalContext* globalCtx);
+void MagicDark_OrbDraw(Actor* thisx, GlobalContext* globalCtx);
 
-void func_80B874E4(Actor* thisx, GlobalContext* globalCtx);
-void func_80B8772C(GlobalContext* globalCtx, f32 a1);
-void func_80B87A18(Actor* thisx, GlobalContext* globalCtx);
+void MagicDark_DiamondUpdate(Actor* thisx, GlobalContext* globalCtx);
+void MagicDark_DimLighting(GlobalContext* globalCtx, f32 intensity);
+void MagicDark_DiamondDraw(Actor* thisx, GlobalContext* globalCtx);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c:20]: I suggest grouping "main" functions together (init/destroy/update/draw) and leaving others like DimLighting apart.

File: src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c (hunk 1 above)
Comment (Sirius902) [reply src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c:115]: Nope, unfortunately it's doing an arithmetic shift here and using `>>=` does a logical shift. For some reason using a temp fixes that.

Comment (Sirius902) [reply src/overlays/actors/ovl_Magic_Dark/z_magic_dark.c]: Okay so apparently `this->primAlpha = (u8)(this->primAlpha >> 1);` works. Very cool.

**Review by Roman971 — APPROVED** [review]

**Review by fig02 — APPROVED** [review]

**Pull request merged**